src/status_im/contexts/wallet/wallet_connect/transactions.cljs

```diff
@@ -15,40 +29,116 @@
                        (not (nil? (get tx tx-key))))
-                (update acc tx-key strip-hex-prefix)
+                (update acc tx-key f)
                 acc))
             tx
             tx-keys)))
 
+;; QUESTION: should we remove the nonce from the original tx so it's always set by status-go, so that
+;; we maintain the order of tx?
+;;
+;; Metamask seems to do so:
+;; https://docs.metamask.io/wallet/how-to/send-transactions#nonce
 (defn- prepare-transaction-for-rpc
   "Formats the transaction and transforms it into a stringified JS object, ready to be passed to an RPC call."
   [tx]
   (-> tx
-      format-tx-hex-values
+      (format-tx-hex-values strip-hex-prefix)
       bean/->js
       (transforms/js-stringify 0)))
 
+(defn beautify-transaction
+  [tx]
+  (let [hex->number #(-> % (subs 2) native-module/hex-to-number)]
+    (-> tx
+        (format-tx-hex-values hex->number)
+        clj->js
+        (js/JSON.stringify nil 2))))
+
+(defn- gwei->hex
+  [gwei]
+  (->> gwei
+       money/gwei->wei
+       native-module/number-to-hex
+       (str "0x")))
+
+(defn- get-max-fee-per-gas-key
+  "Mapping transaction priority (which determines how quickly a tx is processed)
+  to the `suggested-routes` key that should be used for `:maxPriorityFeePerGas`.
+
+  Returns `:high` | `:medium` | `:low`"
+  [tx-priority]
+  (get {:high   :maxFeePerGasHigh
+        :medium :maxFeePerGasMedium
+        :low    :maxFeePerGasLow}
+       tx-priority))
+
+(defn- dynamic-fee-tx?
+  "Checks if a transaction has dynamic fees (EIP1559)"
+  [tx]
+  (every? tx [:maxFeePerGas :maxPriorityFeePerGas]))
+
+(defn- tx->eip1559-tx
+  "Adds `:maxFeePerGas` and `:maxPriorityFeePerGas` for dynamic fee support (EIP1559) and
+  removes `:gasPrice`, if the chain supports EIP1559 and the transaction doesn't already
+  have dynamic fees."
+  [tx suggested-fees tx-priority]
+  (if (and (:eip1559Enabled suggested-fees)
+           (not (dynamic-fee-tx? tx)))
+    (let [max-fee-per-gas-key      (get-max-fee-per-gas-key tx-priority)
+          max-fee-per-gas          (-> suggested-fees max-fee-per-gas-key gwei->hex)
+          max-priority-fee-per-gas (-> suggested-fees :maxPriorityFeePerGas gwei->hex)]
+      (-> tx
+          (assoc
+           :maxFeePerGas         max-fee-per-gas
+           :maxPriorityFeePerGas max-priority-fee-per-gas)
+          ;; NOTE: `:gasPrice` is used only for legacy Tx, so we discard it in favor of dynamic fees
+          (dissoc :gasPrice)))
+    tx))
+
+(defn- prepare-transaction-fees
+  "Makes sure the transaction has the correct gas and fees properties"
+  [tx tx-priority suggested-fees]
+  (-> (assoc tx
+             ;; NOTE: `gasLimit` is ignored on status-go when building a transaction
+             ;; (`wallet_buildTransaction`), so we're setting it as the `gas` property
+             :gas
+             (or (:gasLimit tx)
+                 (:gas tx)))
+      (tx->eip1559-tx suggested-fees tx-priority)))
+
+(defn prepare-transaction
```

Comment:
The lack of tests is a bit concerning to me. Integration tests could be quite effective to help WC code evolve well. Somewhat related to this point, do you think we will move more logic out of the client and into status-go? Is the desktop client going to need to re-implement similar things or do you guys think the client code is already at the sweet spot of logic in client vs server?

Reply:
I think some of the logic will eventually move to status-go, as lots of it are shared. Desktop is doing most of these things as well as currently we use status-go as a persistance layer mostly with minimal logic. Then we'll build up from there.

On the point of testing, we definitely should add tests for these flows, but due to deadlines test seem to be always at the bottom of priorities. Not ideal, but I made an issue to add [integration tests](https://github.com/status-im/status-mobile/issues/20776) and [malli schemas](https://github.com/status-im/status-mobile/issues/20777) for the edges of our API with wallet-connect and status-go, etc.

Reply:
This is perfectly reasonable given the state of affairs :) Thank you @clauxx